File: backend/app/models/schemas/document.py
"""Document-related schemas."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field
from app.utils.time import serialize_utc_datetime

DocumentIndexStatus = Literal["queued", "processing", "indexed", "failed"]
DocumentLifecycleStatus = Literal[
    "draft",
    "pending_review",
    "approved",
    "published",
    "archived",
]
IndexJobStatus = Literal["queued", "processing", "completed", "partial_failed", "failed"]
UTC_MODEL_CONFIG = ConfigDict(json_encoders={datetime: serialize_utc_datetime})


class DocumentCreate(BaseModel):
    """Create a document from text content."""

    model_config = UTC_MODEL_CONFIG

    title: str = Field(..., description="Document title")
    content: str = Field(..., description="Document content")
    document_type: str | None = Field(None, description="Document type such as txt/md/pdf/docx")
    knowledge_base_id: int | None = Field(None, description="Owning knowledge base id")
    category_id: int | None = Field(None, description="Owning category id")
    source_path: str | None = Field(None, description="Original relative source path")


class DocumentResponse(BaseModel):
    """Document detail response."""

    model_config = ConfigDict(from_attributes=True, json_encoders={datetime: serialize_utc_datetime})

    id: int
    title: str
    content: str
    document_type: str | None = None
    size: int = 0
    version: int = 1
    is_current: bool = True
    is_latest: bool = True
    is_live: bool = False
    knowledge_base_id: int | None = Field(None, description="Owning knowledge base id")
    category_id: int | None = Field(None, description="Owning category id")
    category_name: str | None = Field(None, description="Owning category name")
    source_path: str | None = Field(None, description="Original relative source path")
    status: DocumentLifecycleStatus = "draft"
    published_at: datetime | None = None
    published_by: int | None = None
    reviewed_at: datetime | None = None
    reviewed_by: int | None = None
    index_status: DocumentIndexStatus = "queued"
    index_error: str | None = None
    indexed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class DocumentListItem(BaseModel):
    """Document list item."""

    model_config = ConfigDict(from_attributes=True, json_encoders={datetime: serialize_utc_datetime})

    id: int
    title: str
    document_type: str | None = None
    size: int = 0
    version: int = 1
    is_current: bool = True
    is_latest: bool = True
    is_live: bool = False
    indexed: bool = False
    index_status: DocumentIndexStatus = "queued"
    index_error: str | None = None
    indexed_at: datetime | None = None
    knowledge_base_id: int | None = None
    knowledge_base_name: str | None = None
    category_id: int | None = None
    category_name: str | None = None
    source_path: str | None = None
    status: DocumentLifecycleStatus = "draft"
    published_at: datetime | None = None
    published_by: int | None = None
    reviewed_at: datetime | None = None
    reviewed_by: int | None = None
    created_at: datetime
    updated_at: datetime


class DocumentListResponse(BaseModel):
    """Paginated document list response."""

    model_config = UTC_MODEL_CONFIG

    items: list[DocumentListItem]
    total: int
    page: int
    page_size: int


class DocumentContentUpdate(BaseModel):
    """Update document content request."""

    model_config = UTC_MODEL_CONFIG

    content: str = Field(..., description="Updated document content")


class DocumentStatusUpdateResponse(DocumentResponse):
    """Document response used by publish/review transitions."""


class DocumentStatusActionRequest(BaseModel):
    """Optional note placeholder for status transitions."""

    model_config = UTC_MODEL_CONFIG

    note: str | None = Field(default=None, description="Optional admin note")


class DocumentVersionItem(BaseModel):
    """Document version item."""

    model_config = ConfigDict(from_attributes=True, json_encoders={datetime: serialize_utc_datetime})

    id: int
    title: str
    version: int
    is_latest: bool
    is_current: bool
    is_live: bool
    created_at: datetime


class DocumentVersionsResponse(BaseModel):
    """Document version list response."""

    model_config = UTC_MODEL_CONFIG

    items: list[DocumentVersionItem]


class ActiveIndexingJob(BaseModel):
    """Compact progress summary for one active indexing job."""

    model_config = UTC_MODEL_CONFIG

    job_id: int
    title: str
    job_type: str
    job_status: IndexJobStatus
    knowledge_base_id: int | None = None
    knowledge_base_name: str | None = None
    queued: int
    processing: int
    indexed: int
    failed: int
    total: int
    created_at: datetime
    updated_at: datetime | None = None
    progress_percent: int


class FailedIndexingItem(BaseModel):
    """Recent failed indexing document entry for the task panel."""

    model_config = UTC_MODEL_CONFIG

    job_id: int
    document_id: int
    title: str
    job_title: str
    knowledge_base_id: int | None = None
    knowledge_base_name: str | None = None
    index_error: str | None = None
    updated_at: datetime


class IndexingPanelSummaryResponse(BaseModel):
    """Aggregated task-panel payload for document indexing."""

    model_config = UTC_MODEL_CONFIG

    queued: int
    processing: int
    indexed: int
    failed: int
    total: int
    has_active: bool
    active_job_count: int
    active_jobs: list[ActiveIndexingJob]
    recent_failed: list[FailedIndexingItem]

File: frontend/src/shared/types/document.ts
export type DocumentIndexStatus = "queued" | "processing" | "indexed" | "failed";
export type GraphIndexStatus = "skipped" | "queued" | "processing" | "finalizing" | "indexed" | "failed";
export type DocumentParseStatus = "queued" | "processing" | "parsed" | "failed";
export type DocumentLifecycleStatus = "draft" | "pending_review" | "published" | "archived";

export interface DocumentSummary {
  id: number;
  title: string;
  document_type: string | null;
  size: number;
  version: number;
  is_current: boolean;
  is_latest: boolean;
  is_live: boolean;
  indexed: boolean;
  parse_status: DocumentParseStatus;
  parse_error: string | null;
  parsed_at: string | null;
  index_status: DocumentIndexStatus;
  index_error: string | null;
  indexed_at: string | null;
  graph_index_status: GraphIndexStatus;
  graph_index_error: string | null;
  graph_indexed_at: string | null;
  knowledge_base_id: number | null;
  knowledge_base_name: string | null;
  category_id: number | null;
  category_name: string | null;
  source_path: string | null;
  source_storage_provider: string | null;
  source_bucket_name: string | null;
  source_object_key: string | null;
  source_file_name: string | null;
  source_file_size: number | null;
  source_content_type: string | null;
  source_etag: string | null;
  status: DocumentLifecycleStatus;
  published_at: string | null;
  published_by: number | null;
  reviewed_at: string | null;
  reviewed_by: number | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentDetail extends DocumentSummary {
  content: string;
}

export interface DocumentChunkSummary {
  id: number;
  document_id: number;
  chunk_kind: string;
  parent_chunk_id: number | null;
  chunk_index: number;
  prev_chunk_id: number | null;
  next_chunk_id: number | null;
  section_path: string | null;
  block_types: string[];
  start_offset: number;
  end_offset: number;
  content: string;
  search_text: string;
  created_at: string;
  updated_at: string;
}

export interface DocumentListResponse {
  items: DocumentSummary[];
  total: number;
  page: number;
  page_size: number;
  status_counts: Record<string, number>;
}

export interface DocumentChunksResponse {
  items: DocumentChunkSummary[];
  total: number;
}

export interface DocumentQueueResponse {
  message: string;
  queued: number;
}

export interface BatchDocumentActionFailure {
  document_id: number;
  detail: string;
}

export interface BatchDocumentActionResponse {
  action: string;
  requested_count: number;
  succeeded_count: number;
  failed_count: number;
  succeeded_ids: number[];
  failures: BatchDocumentActionFailure[];
}

export interface DocumentUploadInitRequest {
  filename: string;
  file_size: number;
  content_type?: string | null;
  knowledge_base_id: number;
  category_id?: number | null;
  source_path?: string | null;
}

export interface DocumentUploadInitResponse {
  upload_session_id: number;
  provider: string;
  method: string;
  bucket_name: string;
  object_key: string;
  upload_url: string;
  expires_at: string;
  form_fields: Record<string, string>;
}
